### PR TITLE
Rewards Pallet

### LINKED ISSUE #594: Parachain Rewards Pallet
### Introduction
Polkadex secured a para chain slot on Polkadot with 973k DOT ($13.8M) contributed to the Polkadex Crowdloan. All the participants who loaned their DOT tokens will receive PDEX tokens. 25% of PDEX tokens receive as rewards can be claimed by participants immediately while the rest of the 75% will linearly unlock. The treasury account will be a pallet account from which rewards funds will be transferred to participants. 

### Flow Diagram
reward_flow

### Storage Required

**InitializeRewards Map**
Key: reward_id
value: (start_block, end_block, initial_percentage)

**Distributor Map**
Key 1: reward_id
Key 2: account_id
value: (total_reward_amount, claim_amount, is_initial_rewards_claimed, is_initialized, lock_id, last_block_rewards_claim)

### Extrinsic Required

1) `create_reward_cycle`

The extrinsic will be used to start a new reward cycle and can only be called by governance. 
`Parameters`
origin: The sudo/governance address
start_block: The block from where we will start releasing rewards
end_block: By the end_block all rewards should have been released. 
reward_id: The id for which vesting needs to be set. 

----------------------------------------------------

2) `initialize_claim_rewards`
The extrinsic will transfer and lock users' rewards in the user account. It will get user reward_info from the constant hashmap. 
`Parameters`
origin: The users address
reward_id: The id for which vesting needs to be set. 

----------------------------------------------------

3) `claim`
The user will use the extrinsic to claim rewards
`Parameters`
origin: The users address
reward_id: The id for which vesting needs to be set. 

----------------------------------------------------

ToDos
- [x] Design Extrinsics
- [x] Design mocking environment for test cases
- [x] Test Cases

### PR DESCRIPTION
**Pull request by nakul1010**
Closes issue #594.

### CODE REVIEW

File: pallets/rewards/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+[package]
+name = "rewards"
+version = "1.0.0"
+edition = "2018"
```
Comment (serhii-temchenko) [inline pallets/rewards/Cargo.toml]: Please change to `2021`

File: pallets/rewards/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+authors = ['Substrate DevHub <https://github.com/substrate-developer-hub>']
+repository = 'https://github.com/substrate-developer-hub/recipes'
```
Comment (serhii-temchenko) [inline pallets/rewards/Cargo.toml]: I assume this should be replaced with:
```
repository = "https://github.com/Polkadex-Substrate/Polkadex/"
```
like with other pallets.

File: pallets/rewards/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,54 @@
+[package]
+name = "rewards"
+version = "1.0.0"
+edition = "2018"
+authors = ['Substrate DevHub <https://github.com/substrate-developer-hub>']
```
Comment (serhii-temchenko) [inline pallets/rewards/Cargo.toml]: Shouldn't this be like [here](https://github.com/Polkadex-Substrate/Polkadex/blob/67cbe958c9adb706dd3aeb92a68b3564ba209cf2/pallets/asset-handler/Cargo.toml#L2)?

File: pallets/rewards/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+	// method.
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		//ToDo: on_intialize would be added to clear events, to clear storage if everyone has been
+		// donated
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Please add issue ref here

File: pallets/rewards/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// The extrinsic will be used to start a new reward cycle
+		/// Parameters
+		/// origin: The donor who wants to start the reward cycle
+		/// start_block: The block from which reward distribution will start
+		/// end_block: The block at which last rewards will be distributed
+		/// initial_percentage: The percentage of rewards that can be claimed at start block
+		/// reward_id: The reward id
+		#[pallet::weight(10_000)]
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs:120]: Didn't we need banchmarks?

File: pallets/rewards/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// The extrinsic will be used to start a new reward cycle
+		/// Parameters
+		/// origin: The donor who wants to start the reward cycle
+		/// start_block: The block from which reward distribution will start
+		/// end_block: The block at which last rewards will be distributed
+		/// initial_percentage: The percentage of rewards that can be claimed at start block
+		/// reward_id: The reward id
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Please use common syntax for dockblocks like [here](https://github.com/Polkadex-Substrate/Polkadex/blob/67cbe958c9adb706dd3aeb92a68b3564ba209cf2/pallets/asset-handler/src/lib.rs#L267-L273)

File: pallets/rewards/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+			origin: OriginFor<T>,
+			start_block: T::BlockNumber,
+			end_block: T::BlockNumber,
+			initial_percentage: u32,
+			reward_id: u32,
+		) -> DispatchResult {
+			//check to ensure governance
+			T::GovernanceOrigin::ensure_origin(origin.clone())?;
+
+			//check to ensure no dupicate id gets added
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Typo

File: pallets/rewards/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+		/// reward id doesn't correctly map to donor
+		IncorrectDonorAccount,
+		/// The reward Id is not register
+		RewardIdNotRegister,
+		/// User not eligible for the reward
+		UserNotEligible,
+		/// Transfer of funds failed
+		TransferFailed,
+		/// Amount to low to reedeem
+		AmountToLowtoRedeem,
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Please rename to `AmountToLowToRedeem`

File: pallets/rewards/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+		/// User needs to intialize first before claiming rewards
+		UserHasNotIntializeClaimRewards,
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Typo `Intialize`

File: pallets/rewards/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+			reward_id: u32,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The id has already been taken
+		DuplicateId,
+		/// start block should be smaller than end block
+		InvalidParameter,
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: I'd suggest rename to something more specific like `InvalidBlocksRange` since now it's to common.

File: pallets/rewards/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+
+			//ensure start block is greater than min
+			ensure!(
+				initial_percentage <= 100 && initial_percentage > 0,
+				Error::<T>::InvalidParameter
+			);
+
+			let reward_info = RewardInfo { start_block, end_block, initial_percentage };
+
+			//inserting rewards info into storage
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Typo

File: pallets/rewards/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,546 @@
...
+			let difference_between_start_and_current_block = start_block
+				.saturated_into::<u128>()
+				.saturating_sub(<frame_system::Pallet<T>>::block_number().saturated_into::<u128>());
+			ensure!(
+				difference_between_start_and_current_block >
+					MIN_DIFFERENCE_BETWEEN_START_AND_END_BLOCK,
+				Error::<T>::InvalidParameter
+			);
+
+			//ensure start block is greater than min
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Not responding to the real assertion subject. Plus should be added assertion regarding this comment?

File: pallets/rewards/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,546 @@
...
+			ensure!(
+				initial_percentage <= 100 && initial_percentage > 0,
+				Error::<T>::InvalidParameter
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: For this assertion mby `InvalidInitialProcentage` error variant could be added?

File: pallets/rewards/src/tests.rs [hunk 14]
```diff
@@ -0,0 +1,1105 @@
...
+}
+
+#[test]
+fn create_reward_cycle_when_percentage_parameter_is_invalid() {
+	new_test_ext().execute_with(|| {
+		let (start_block, end_block, _, reward_id) = get_parameters_for_reward_cycle();
+		assert_noop!(
+			Rewards::create_reward_cycle(Origin::root(), end_block, start_block, 101, reward_id),
+			Error::<Test>::InvalidParameter
+		);
```
Comment (serhii-temchenko) [inline pallets/rewards/src/tests.rs]: Here asserted only one case, it would be nice to have this assertion for `0` as well.

File: pallets/rewards/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,546 @@
...
+			//ensure that difference between start of vesting period - current block is greater
+			// than min difference
+			let difference_between_start_and_current_block = start_block
+				.saturated_into::<u128>()
+				.saturating_sub(<frame_system::Pallet<T>>::block_number().saturated_into::<u128>());
+			ensure!(
+				difference_between_start_and_current_block >
+					MIN_DIFFERENCE_BETWEEN_START_AND_END_BLOCK,
+				Error::<T>::InvalidParameter
+			);
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: I can't find tests for this part. Point me please if its there and if not - please add.

File: pallets/rewards/src/mock.rs [hunk 16]
```diff
@@ -0,0 +1,163 @@
+use crate::*;
+// use crate::mock::sp_api_hidden_includes_construct_runtime::hidden_include::traits::{
+// 	OnFinalize, OnInitialize,
+// };
```
Comment (serhii-temchenko) [inline pallets/rewards/src/mock.rs]: Please remove

File: pallets/rewards/src/mock.rs [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+	type WeightInfo = ();
+}
+
+// impl Config for Test {
+// 	type Event = Event;
+// 	type NativeCurrency = Balances;
+// 	type OtherAssets = Assets;
+// 	type Public = <Signature as sp_runtime::traits::Verify>::Signer;
+// 	type Signature = Signature;
+// }
```
Comment (serhii-temchenko) [inline pallets/rewards/src/mock.rs]: Please remove

File: pallets/rewards/types.json [hunk 18]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (serhii-temchenko) [inline pallets/rewards/types.json]: What is this file about?

File: pallets/rewards/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,546 @@
...
+			Ok(())
+		}
+
+		///The extrinsic will add beneficiaries for particular reward id
+		/// Parameters,
+		/// origin: The donor for the particular reward id
+		/// id: Reward id
+		/// conversion_factor: The conversion factor from dot to pdex
+		/// beneficiaries: The account id who can claim the reward & the amount in dot contributed
+		/// base 10^12 u128: the value provide here considers 1 unit = 10^12
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Same

File: pallets/rewards/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,546 @@
...
+				//will not occur since we are already ensuring it above, still sanity check
+				return Err(Error::<T>::RewardIdNotRegister.into())
+			}
+			Ok(())
+		}
+
+		///The extrinsic will transfer and lock users rewards in the users account
+		/// Parameters,
+		/// origin: The users address which has been mapped to reward id
+		/// id: Reward id
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Same

File: pallets/rewards/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,546 @@
...
+					Err(Error::<T>::UserNotEligible)
+				}
+			})?;
+			Self::deposit_event(Event::UserUnlockedReward { user, reward_id });
+			Ok(())
+		}
+
+		/// The user will use the extrinsic to claim rewards
+		/// origin: The users address which has been mapped to reward id
+		/// id: The reward id
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Same

File: pallets/rewards/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,546 @@
...
+		RewardInfoForAccount<T>,
+		OptionQuery,
+	>;
+}
+
+// The main implementation block for the pallet. Functions here fall into three broad
+// categories:
+// - Public interface. These are functions that are `pub` and generally fall into inspector
+// functions that do not write to storage and operation functions that do.
+// - Private functions. These are your usual private utilities unavailable to other pallets.
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Please remove

File: pallets/rewards/src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,546 @@
...
+impl<T: Config> Pallet<T> {
+	fn get_pallet_account() -> T::AccountId {
+		T::PalletId::get().into_account_truncating()
+	}
+
+	//The followling function will be used by claim extrinsic to tranfer balance from donor to
+	// beneficiary
+
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Remove please empty line

File: pallets/rewards/src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,546 @@
...
+				}
+			})?;
+
+			Ok(())
+		}
+	}
+
+	/// Events are a simple means of reporting specific conditions and
+	/// circumstances that have happened that users, Dapps and/or chain explorers would find
+	/// interesting and otherwise difficult to detect.
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Please remove

File: pallets/rewards/src/lib.rs [hunk 25]
```diff
@@ -0,0 +1,546 @@
...
+				if let Some(user_reward_info) = user_reward_info {
+					// only unlock reward if current block greater than or equal to the starting
+					// block of reward
+					if let Some(reward_info) = <InitializeRewards<T>>::get(reward_id) {
+						ensure!(
+							reward_info.start_block.saturated_into::<u128>() <=
+								<frame_system::Pallet<T>>::block_number()
+									.saturated_into::<u128>(),
+							Error::<T>::RewardsCannotBeUnlockYet
+						);
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs:275]: This assertion is not covered in tests

File: pallets/rewards/src/lib.rs [hunk 26]
```diff
@@ -0,0 +1,546 @@
...
+					}
+
+					//check if user already unlocked the rewards
+					ensure!(!user_reward_info.is_initialized, Error::<T>::RewardsAlreadyUnlocked);
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs:279]: This assertion is not covered in tests

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/rewards/src/lib.rs [hunk 27]
```diff
@@ -0,0 +1,546 @@
...
+						ensure!(
+							user_reward_info.claim_amount.saturated_into::<u128>() !=
+								user_reward_info.total_reward_amount.saturated_into::<u128>(),
+							Error::<T>::AllRewardsAlreadyClaimed
+						);
+
+						let mut rewards_claimable: u128 = 0_u128.saturated_into();
+
+						//calculate the intial rewards that can be claimed
+						let initial_rewards_claimed = user_reward_info
```
Comment (mcdonagg) [inline pallets/rewards/src/lib.rs]: Should this only be calculated if the user has not already claimed the rewards? Instead of calculating then checking?

File: pallets/rewards/src/lib.rs [hunk 28]
```diff
@@ -0,0 +1,546 @@
...
+							min(current_block_no, reward_info.end_block.saturated_into::<u128>())
+								.saturating_sub(last_reward_claimed_block_no);
+						let crowdloan_period = reward_info
+							.end_block
+							.saturated_into::<u128>()
+							.saturating_sub(reward_info.start_block.saturated_into::<u128>());
+
+						//calculate custom factor for the user
+						// Formula = (total_rewards - intial_rewards_claimed) / crowloan_period
+						let factor = user_reward_info
```
Comment (mcdonagg) [inline pallets/rewards/src/lib.rs]: Can this factor be calculated and stored on the user_reward_info while initializing, so it only needs to be calculated once instead of on every claim?

File: pallets/rewards/src/lib.rs (hunk 4 above)
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: not required anymore

File: pallets/rewards/src/lib.rs (hunk 5 above)
Comment (nakul1010) [reply pallets/rewards/src/lib.rs:120]: issue #603 will fix that.

File: pallets/rewards/src/lib.rs (hunk 11 above)
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: reward

File: pallets/rewards/types.json (hunk 18 above)
Comment (nakul1010) [reply pallets/rewards/types.json]: removed

File: pallets/rewards/src/lib.rs (hunk 15 above)
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: `create_reward_cycle_when_block_range_is_invalid` added

File: pallets/rewards/src/lib.rs (hunk 25 above)
Comment (nakul1010) [reply pallets/rewards/src/lib.rs:275]: `initialize_claim_rewards_when_vesting_period_not_started` test case added

File: pallets/rewards/src/lib.rs (hunk 26 above)
Comment (nakul1010) [reply pallets/rewards/src/lib.rs:279]: Added it in `claim_rewards_at_start_block` test case

File: pallets/rewards/src/lib.rs (hunk 11 above)
Comment (serhii-temchenko) [reply pallets/rewards/src/lib.rs]: sorry my bad, misread 'info ' and thought that `into` repeated twice....

File: pallets/rewards/src/lib.rs (hunk 27 above)
Comment (nakul1010) [reply pallets/rewards/src/lib.rs:338]: it is also used for calculating the factor so it doesn't matter if u calculate it afterward.

**Review by serhii-temchenko — DISMISSED** [review]

File: pallets/rewards/src/lib.rs (hunk 28 above)
Comment (mcdonagg) [reply pallets/rewards/src/lib.rs]: Thank you for implementing.

File: pallets/rewards/src/lib.rs (hunk 27 above)
Comment (mcdonagg) [reply pallets/rewards/src/lib.rs]: I like your implementation, that looks great

File: pallets/rewards/Cargo.toml [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+parity-scale-codec = { workspace = true, default-features = false }
+log = { workspace = true, default-features = false }
+scale-info = { workspace = true, default-features = false, features = ["derive"] }
+frame-support = { default-features = false, workspace = true }
+frame-system = { default-features = false, workspace = true }
+sp-runtime = { default-features = false, workspace = true }
+sp-std = { workspace = true, default-features = false }
+sp-core = { workspace = true, default-features = false }
+polkadex-primitives = { workspace = true, default-features = false }
+ias-verify = { workspace = true, default-features = false }
```
Comment (Gauthamastro) [inline pallets/rewards/Cargo.toml]: Why is this here?

Comment (Gauthamastro) [reply pallets/rewards/Cargo.toml]: Clean cargo toml

File: runtime/src/lib.rs [hunk 30]
```diff
@@ -1271,6 +1271,20 @@ impl pallet_ocex_lmp::Config for Runtime {
...
 
+parameter_types! {
+	pub const RewardsPalletId: PalletId = PalletId(*b"REWARDSQ");
+}
+
+impl rewards::Config for Runtime {
+	type Event = Event;
+	type PalletId = RewardsPalletId;
+	type NativeCurrency = Balances;
+	type OtherAssets = Assets;
```
Comment (Gauthamastro) [inline runtime/src/lib.rs]: Why do we need OtherAssets?

File: pallets/rewards/src/lib.rs [hunk 31]
```diff
@@ -0,0 +1,545 @@
...
+type BalanceOf<T> =
+	<<T as Config>::NativeCurrency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(test)]
+mod mock;
+
+const UNIT_BALANCE: u128 = 1_000_000_000_000;
```
Comment (Gauthamastro) [inline pallets/rewards/src/lib.rs]: Use this from polkadex-primitives

File: pallets/rewards/src/lib.rs [hunk 32]
```diff
@@ -0,0 +1,545 @@
...
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Balances Pallet
+		type NativeCurrency: Currency<Self::AccountId>
+			+ ReservableCurrency<Self::AccountId>
+			+ LockableCurrency<Self::AccountId>;
+
+		/// Assets Pallet
+		type OtherAssets: Mutate<
```
Comment (Gauthamastro) [inline pallets/rewards/src/lib.rs]: Why?

**Review by Gauthamastro — CHANGES_REQUESTED** [review]

File: pallets/rewards/Cargo.toml (hunk 29 above)
Comment (nakul1010) [reply pallets/rewards/Cargo.toml]: Removed unused dependency.

File: runtime/src/lib.rs (hunk 30 above)
Comment (nakul1010) [reply runtime/src/lib.rs]: Removed otherassets

File: pallets/rewards/src/lib.rs (hunk 32 above)
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: removed

File: pallets/rewards/src/lib.rs [hunk 33]
```diff
@@ -0,0 +1,534 @@
...
+	dispatch::DispatchResult,
+	pallet_prelude::Get,
+	traits::{Currency, ExistenceRequirement, LockIdentifier},
+	BoundedVec,
+};
+
+use pallet_timestamp::{self as timestamp};
+use sp_runtime::traits::{AccountIdConversion, UniqueSaturatedInto};
+use sp_std::prelude::*;
+
```
Comment (35359595) [inline pallets/rewards/src/lib.rs]: Remove empty line

File: pallets/rewards/src/lib.rs [hunk 34]
```diff
@@ -0,0 +1,534 @@
...
+use sp_runtime::SaturatedConversion;
```
Comment (35359595) [inline pallets/rewards/src/lib.rs]: Combine same crate imports into one.

File: pallets/rewards/src/lib.rs [hunk 35]
```diff
@@ -0,0 +1,534 @@
...
+	use frame_support::{
+		pallet_prelude::{OptionQuery, *},
+		traits::{Currency, LockableCurrency, ReservableCurrency},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::*;
+
+	use sp_runtime::traits::{IdentifyAccount, Verify};
+
+	use frame_support::traits::WithdrawReasons;
```
Comment (35359595) [inline pallets/rewards/src/lib.rs]: Clean up imports: join same crate ones, remove empty lines

File: pallets/rewards/src/lib.rs [hunk 36]
```diff
@@ -0,0 +1,534 @@
...
+			origin: OriginFor<T>,
+			reward_id: u32,
+			conversion_factor: BalanceOf<T>,
+			beneficiaries: BoundedVec<
+				(T::AccountId, BalanceOf<T>),
+				polkadex_primitives::ingress::HandleBalanceLimit,
+			>,
+		) -> DispatchResult {
+			//check to ensure governance
+			T::GovernanceOrigin::ensure_origin(origin.clone())?;
```
Comment (35359595) [inline pallets/rewards/src/lib.rs]: Is this clone required?

File: pallets/rewards/src/lib.rs [hunk 37]
```diff
@@ -0,0 +1,534 @@
...
+						//check if user has initialize rewards or not
+						ensure!(
+							user_reward_info.is_initialized,
+							Error::<T>::UserHasNotInitializeClaimRewards
+						);
+
+						//ensure that all rewards are not already claimed
+						ensure!(
+							user_reward_info.claim_amount.saturated_into::<u128>() !=
+								user_reward_info.total_reward_amount.saturated_into::<u128>(),
```
Comment (35359595) [inline pallets/rewards/src/lib.rs]: Bigger is also not equal. Set it to claimed minus current is bigger than curren. Unclaimed must be enough to cover current claim

File: pallets/rewards/src/lib.rs [hunk 38]
```diff
@@ -0,0 +1,534 @@
...
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The id has already been taken
+		DuplicateId,
+		/// invalid block range provided
+		InvalidBlocksRange,
+		///  invalid percentage range
```
Comment (35359595) [inline pallets/rewards/src/lib.rs]: Make it consistent capital letters first and align whitespaces

File: runtime/Cargo.toml [hunk 39]
```diff
@@ -180,6 +183,7 @@ std = [
     "asset-handler/std",
     "chainbridge/std",
     "pallet-ocex-lmp/std",
+    "rewards/std",
```
Comment (35359595) [inline runtime/Cargo.toml:190]: Align with others

**Review by 35359595 — CHANGES_REQUESTED** [review]

File: runtime/Cargo.toml (hunk 39 above)
Comment (35359595) [reply runtime/Cargo.toml:190]: wrong view in browser. resolved

File: pallets/rewards/src/lib.rs (hunk 36 above)
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: not required

File: pallets/rewards/src/lib.rs (hunk 37 above)
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: Fixed in the latest commit

**Comment by serhii-temchenko** [comment]
@nakul1010 please fix CI

File: pallets/rewards/src/lib.rs [hunk 40]
```diff
@@ -0,0 +1,535 @@
...
+						rewards_claimable = rewards_claimable.saturating_add(
+							user_reward_info
+								.factor
+								.saturated_into::<u128>()
+								.saturating_mul(unclaimed_blocks),
+						);
+
+						//ensure total_rewards_claimable - rewards_claimed >= rewards_claimable
+						//sanity check
+						ensure!(
```
Comment (nakul1010) [inline pallets/rewards/src/lib.rs:362]: This condition should never trigger unless there is something wrong in the calculation or governance modifies the storage. I can't just put greater than, as the user will not be able to claim the last claimable reward. 

If a user tries to claim a reward when all rewards are already claimed rewards_claimable will have a value of zero and the `MIN_REWARDS_CLAIMABLE_AMOUNT` ensure will trigger.

Comment (35359595) [reply pallets/rewards/src/lib.rs:362]: >= will do

**Review by 35359595 — DISMISSED** [review]
LGTM

**Review by serhii-temchenko — DISMISSED** [review]

**Review by Gauthamastro — CHANGES_REQUESTED** [review]
This is different from how it is supposed to be built.
